### PR TITLE
Ospi with DMA transfer for stm32U5 devices

### PR DESCRIPTION
**Pull request by FRASTM**
This PR enables the octoSPI NOR flash that is present on the  stm32u585 disco kit

The octospi peripheral are defined by the DTS of the devices.
The boards enable the mx25lm51245 NOR octo-flash controller.
The file overlay is enabling the GPDMA (channel12) for OSPI transfer (This corresponds to the request 41)

It requires https://github.com/zephyrproject-rtos/zephyr/pull/45151

Signed-off-by: Francois Ramu [francois.ramu@st.com](mailto:francois.ramu@st.com)

### CODE REVIEW

**Comment by FRASTM** [comment]
$ west build -p auto -b b_u585i_iot02a samples/drivers/spi_flash
```
*** Booting Zephyr OS build zephyr-v3.1.0-4139-g2a2ee724af89  ***

ospi-nor-flash@0 SPI flash testing
==========================

Test 1: Flash erase
Flash erase succeeded!

Test 2: Flash write
Attempting to write 4 bytes
Data read matches data written. Good!!
```

**Comment by stephanosio** [comment]
Rebased to fix CI failure

**Comment by FRASTM** [comment]
rebase on 9a9fc2b8a5ddf81e44330d9cbebaeba360651e68
fixing conflicts

File: drivers/flash/Kconfig.stm32_ospi [hunk 1]
```diff
@@ -20,5 +20,6 @@ config FLASH_STM32_OSPI
 	select DMA if $(DT_STM32_OCTOSPI_1_HAS_DMA) || $(DT_STM32_OCTOSPI_2_HAS_DMA)
 	select USE_STM32_HAL_DMA if $(DT_STM32_OCTOSPI_1_HAS_DMA) || \
 					$(DT_STM32_OCTOSPI_2_HAS_DMA)
+	select USE_STM32_HAL_DMA_EX if SOC_SERIES_STM32U5X && ($(DT_STM32_OCTOSPI_1_HAS_DMA) || $(DT_STM32_OCTOSPI_2_HAS_DMA))
```
Comment (erwango) [inline drivers/flash/Kconfig.stm32_ospi:23]: Fix long line

File: drivers/flash/flash_stm32_ospi.c [hunk 2]
```diff
@@ -50,6 +51,27 @@ LOG_MODULE_REGISTER(flash_stm32_ospi, CONFIG_FLASH_LOG_LEVEL);
 #define SPI_NOR_WRITEOC_NONE 0xFF
 
 #if STM32_OSPI_USE_DMA
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:54]: Please use compatible

File: drivers/flash/flash_stm32_ospi.c [hunk 3]
```diff
@@ -1691,29 +1720,55 @@ static int flash_stm32_ospi_init(const struct device *dev)
 
 	int index = find_lsb_set(dma_cfg.source_data_size) - 1;
 
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:1723]: Please use compatible

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/flash/flash_stm32_ospi.c (hunk 2 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: using CONFIG_DT_HAS_ST_STM32U5_DMA_ENABLED

File: drivers/flash/flash_stm32_ospi.c (hunk 3 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: using CONFIG_DT_HAS_ST_STM32U5_DMA_ENABLED

File: drivers/flash/Kconfig.stm32_ospi (hunk 1 above)
Comment (FRASTM) [reply drivers/flash/Kconfig.stm32_ospi]: reduce line length

**Comment by FRASTM** [comment]
with  `CONFIG_DT_HAS_ST_STM32U5_DMA_ENABLED` instead of `defined(CONFIG_SOC_SERIES_STM32U5X)`
OSPI and DMA and stm32u5 serie

File: drivers/flash/flash_stm32_ospi.c [hunk 4]
```diff
@@ -50,6 +51,27 @@ LOG_MODULE_REGISTER(flash_stm32_ospi, CONFIG_FLASH_LOG_LEVEL);
 #define SPI_NOR_WRITEOC_NONE 0xFF
 
 #if STM32_OSPI_USE_DMA
+#if CONFIG_DT_HAS_ST_STM32U5_DMA_ENABLED
```
Comment (erwango) [inline drivers/flash/flash_stm32_ospi.c:54]: We can't use this symbol. Its usage is restricted.
We should use DMA_STM32U5

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/flash/flash_stm32_ospi.c (hunk 4 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c:54]: using CONFIG_DMA_STM32U5

**Comment by FRASTM** [comment]
with  CONFIG_DMA_STM32U5

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Pull request merged**